### ISSUE TITLE
NaNs in torch.linalg.eigh backward

### ISSUE THREAD

**Issue by Lmy0217** [pytorch/pytorch#76531]
### 🐛 Describe the bug

```python
import torch

K1 = torch.tensor([[-0.3329024952074725,  0.0,  0.0,  0.0],
        [ 0.0019572051122251537, -0.3244421706868837,  0.00000,  0.00000],
        [-0.0016241708944436939, -0.007378259691806472, -0.3272105439699563,  0.00000],
        [ 0.023828483586765488,  0.10824768536769169, -0.08982846962075508,  0.9845552098643126]], device='cuda:0', dtype=torch.float64, requires_grad=True)

g1 = torch.tensor([[ 0.0,  0.0,  0.0,  7.273988740053028e-05],
        [ 0.0,  0.0,  0.0, -0.00011222996545257047],
        [ 0.0,  0.0,  0.0, -0.0005098145338706672],
        [ 0.0,  0.0,  0.0,  0.00042305406532250345]], device='cuda:0', dtype=torch.float64)

print(torch.linalg.eigh(K1)[1].grad_fn(None, g1))
# tensor([[nan, nan, nan, nan],
#         [nan, nan, nan, nan],
#         [nan, nan, nan, nan],
#         [nan, nan, nan, nan]], device='cuda:0', dtype=torch.float64,
#        grad_fn=<MmBackward0>)
```

### Versions

PyTorch version: 1.11.0+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-11ubuntu0~18.04.1) 9.3.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.8.5 (default, Sep  4 2020, 07:30:14)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-80-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.1.74
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090
GPU 3: NVIDIA GeForce RTX 3090

Nvidia driver version: 470.57.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] nfnets-pytorch==0.1.3
[pip3] numpy==1.22.3
[pip3] numpydoc==1.1.0
[pip3] performer-pytorch==0.15.0
[pip3] pytorch-fast-transformers==0.3.0
[pip3] reformer-pytorch==1.2.4
[pip3] robust-loss-pytorch==0.0.2
[pip3] torch==1.11.0+cu113
[pip3] torch-dct==0.1.5
[pip3] torchaudio==0.11.0+cu113
[pip3] torchio==0.18.75
[pip3] torchvision==0.12.0+cu113
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.2.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] nfnets-pytorch            0.1.3                    pypi_0    pypi
[conda] numpy                     1.22.3                   pypi_0    pypi
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] performer-pytorch         0.15.0                   pypi_0    pypi
[conda] pytorch-fast-transformers 0.3.0                    pypi_0    pypi
[conda] reformer-pytorch          1.2.4                    pypi_0    pypi
[conda] torch                     1.11.0+cu113             pypi_0    pypi
[conda] torch-dct                 0.1.5                    pypi_0    pypi
[conda] torchaudio                0.11.0+cu113             pypi_0    pypi
[conda] torchio                   0.18.75                  pypi_0    pypi
[conda] torchvision               0.12.0+cu113             pypi_0    pypi

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

**Comment by nikitaved** [comment]
If you get to look at the eigenvalues of this matrix, there are 3 eigenvalues which are all approximately equal to `-0.33`. Have a closer look at the second warning in the documentation: https://pytorch.org/docs/stable/generated/torch.linalg.eigh.html?highlight=eigh#torch.linalg.eigh. Unfortunately, there is not much we can do here, unless your objective function does not depend on the eigenspace but on eigenvalues. In such a case it should be well-defined.

**Comment by Lezcano** [comment]
Closing as this is expected (the function is not differentiable at the given matrix).
